### PR TITLE
Add segment merged result cache for broker

### PR DESCRIPTION
**Pull request by yuanlihan**
### Description
This PR adds a new cache feature for broker.

### Motivation
Currently, the broker can't provide a caching capability for relatively fixed queries need to scan both historical and realtime segments and with time range filter, like the day so far or this week/month/year so far.
It would be useful to have a cache feature to cache merged results of historical segments on Broker. It could be expected as a better cache option on Broker for the following use cases.
|                      | Case 1: can work when using group by v2 engine | Case 2: can work when scanning historical segments only | Case 3: can work when scanning both historical and realtime segments | Case 4: be efficient to cache large number of segments |
| -------------------- | ---------------------------------------------- | ------------------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------ |
| Segment Level cache  | &cross;                                        | &check;                                                 | &check;                                                      | &cross;                                                |
| Result level cache   | &cross;                                        | &check;                                                 | &cross;                                                      | &check;                                                |
| **New segment merged cache** | &check;                                        | &check;                                                 | &check;                                                      | &check;                                                |

### Proposed changes
|Property|Possible Values|Description|Default|
|--------|---------------|-----------|-------|
|`druid.broker.cache.useSegmentMergedResultCache`|true, false|Enable segment merged result cache on the Broker. This cache option can be useful for queries need to scan both historical and realtime segments and with time range filter, like the day so far or this week/month/year so far. It can't work together with segment level cache(see `druid.broker.cache.useCache` and `druid.broker.cache.populateCache`) on Broker. If both segment merged result cache and segment level cache are enabled at the same time, then only segment merged result cache will work.|false|
|`druid.broker.cache.populateSegmentMergedResultCache`|true, false|Populate segment merged result cache on the Broker. See also `druid.broker.cache.useSegmentMergedResultCache`.|false|

### Rationale
As we know that the segment level cache on broker can't work for group by v2 engine and is not very efficient for large number of segments. And the result level cache also can't work for queries need to scan realtime segments. The segment merged result cache tries to solve the limitation of the existing two cache options on Broker. The segment merged result caching feature works like
image

### Operational impact

1. recommend to increase the `cache.maxEntrySize` on broker
2. the `*/hits` and `*/hitRate` metrics might seem incorrect from query caching hitting view as we need to check extra info from the cache entry to determine whether to accept the entry. The result level cache has same issue. 

### Test plan
Below is the benchmark result(`useParallelMergePool=true` by default)  or view [graph of benchmark result](https://jmh.morethan.io/?gist=5bbb3e650304496f4a566cac8eb4915f)
```text
Benchmark                                          (cacheOption)  (numHistoircals)  (numPeons)  (numSegmentsPerHistorial)  (numSegmentsPerPeon)  (queryGranularity)  (round)  (rowsPerSegment)  Mode  Cnt      Score      Error  Units
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           0                          5                     1              minute        1             20000  avgt    5   7973.685 ± 1494.451  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           0                          5                     1              minute        3             20000  avgt    5  21964.523 ±  410.604  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           1                          5                     1              minute        1             20000  avgt    5   7630.985 ±  132.387  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           1                          5                     1              minute        3             20000  avgt    5  23981.028 ± 4070.059  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           0                          5                     1              minute        1             20000  avgt    5  16569.784 ±  689.911  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           0                          5                     1              minute        3             20000  avgt    5  50170.979 ± 4558.294  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           1                          5                     1              minute        1             20000  avgt    5  16194.900 ± 2549.430  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           1                          5                     1              minute        3             20000  avgt    5  36463.924 ± 1551.288  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           0                          5                     1              minute        1             20000  avgt    5   6301.880 ±   95.760  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           0                          5                     1              minute        3             20000  avgt    5  18967.467 ±  182.751  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           1                          5                     1              minute        1             20000  avgt    5   8347.986 ±  142.536  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           1                          5                     1              minute        3             20000  avgt    5  26691.552 ±  244.504  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           0                          5                     1              minute        1             20000  avgt    5  14577.487 ±  201.302  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           0                          5                     1              minute        3             20000  avgt    5  31139.173 ±  250.591  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           1                          5                     1              minute        1             20000  avgt    5  14412.902 ±  121.302  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           1                          5                     1              minute        3             20000  avgt    5  41348.599 ±  375.229  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           0                          5                     1              minute        1             20000  avgt    5   2154.803 ±  785.275  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           0                          5                     1              minute        3             20000  avgt    5   5379.564 ±  256.809  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           1                          5                     1              minute        1             20000  avgt    5   2662.320 ±   97.894  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           1                          5                     1              minute        3             20000  avgt    5  10078.134 ±   85.502  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           0                          5                     1              minute        1             20000  avgt    5  14904.827 ± 1738.628  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           0                          5                     1              minute        3             20000  avgt    5  17122.941 ± 1003.016  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           1                          5                     1              minute        1             20000  avgt    5  19328.719 ±  270.126  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           1                          5                     1              minute        3             20000  avgt    5  18077.747 ±  467.460  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           0                          5                     1              minute        1             20000  avgt    5   9535.534 ±  809.587  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           0                          5                     1              minute        3             20000  avgt    5  25663.662 ±   90.529  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           1                          5                     1              minute        1             20000  avgt    5   8978.321 ±  154.007  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           1                          5                     1              minute        3             20000  avgt    5  26570.464 ±  118.909  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           0                          5                     1              minute        1             20000  avgt    5  15443.262 ±  229.761  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           0                          5                     1              minute        3             20000  avgt    5  44848.647 ± 1534.248  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           1                          5                     1              minute        1             20000  avgt    5  15134.711 ±  227.012  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           1                          5                     1              minute        3             20000  avgt    5  44943.205 ±  370.755  ms/op
```

Below is the benchmark result(disable `useParallelMergePool`)  or view [graph of benchmark result](https://jmh.morethan.io/?gist=091a169363b68b05a82eaf70b26cbcdc)
```text
Benchmark                                          (cacheOption)  (numHistoircals)  (numPeons)  (numSegmentsPerHistorial)  (numSegmentsPerPeon)  (queryGranularity)  (round)  (rowsPerSegment)  Mode  Cnt      Score      Error  Units
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           0                          5                     1              minute        1             20000  avgt    5   9917.528 ± 2933.381  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           0                          5                     1              minute        3             20000  avgt    5  22773.857 ±  161.082  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           1                          5                     1              minute        1             20000  avgt    5   8413.935 ±   75.407  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                 5           1                          5                     1              minute        3             20000  avgt    5  24826.290 ±  191.913  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           0                          5                     1              minute        1             20000  avgt    5  19588.804 ± 2544.812  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           0                          5                     1              minute        3             20000  avgt    5  44042.234 ± 2783.275  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           1                          5                     1              minute        1             20000  avgt    5  19528.356 ±  938.254  ms/op
CachingClusteredClientCachingBenchmark.runQueries        segment                10           1                          5                     1              minute        3             20000  avgt    5  45506.638 ±  443.176  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           0                          5                     1              minute        1             20000  avgt    5   7645.671 ±  106.838  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           0                          5                     1              minute        3             20000  avgt    5  21489.430 ±  464.426  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           1                          5                     1              minute        1             20000  avgt    5   9285.202 ± 1238.785  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                 5           1                          5                     1              minute        3             20000  avgt    5  30569.659 ±  248.147  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           0                          5                     1              minute        1             20000  avgt    5  16903.978 ±  369.545  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           0                          5                     1              minute        3             20000  avgt    5  37956.136 ±  338.327  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           1                          5                     1              minute        1             20000  avgt    5  20355.786 ±  360.331  ms/op
CachingClusteredClientCachingBenchmark.runQueries         result                10           1                          5                     1              minute        3             20000  avgt    5  63560.170 ±  795.199  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           0                          5                     1              minute        1             20000  avgt    5   3532.208 ±   89.560  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           0                          5                     1              minute        3             20000  avgt    5  10776.855 ±  144.780  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           1                          5                     1              minute        1             20000  avgt    5   5350.147 ±   42.341  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                 5           1                          5                     1              minute        3             20000  avgt    5  12626.949 ± 1081.221  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           0                          5                     1              minute        1             20000  avgt    5  19181.467 ±  147.430  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           0                          5                     1              minute        3             20000  avgt    5  22088.464 ±  267.739  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           1                          5                     1              minute        1             20000  avgt    5  18175.182 ±  226.354  ms/op
CachingClusteredClientCachingBenchmark.runQueries  segmentMerged                10           1                          5                     1              minute        3             20000  avgt    5  23285.225 ±  401.285  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           0                          5                     1              minute        1             20000  avgt    5   9021.147 ±   85.542  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           0                          5                     1              minute        3             20000  avgt    5  26804.468 ± 5661.785  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           1                          5                     1              minute        1             20000  avgt    5   9474.065 ±   87.597  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                 5           1                          5                     1              minute        3             20000  avgt    5  28422.492 ±  243.659  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           0                          5                     1              minute        1             20000  avgt    5  15796.851 ±  104.173  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           0                          5                     1              minute        3             20000  avgt    5  52883.206 ±  332.724  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           1                          5                     1              minute        1             20000  avgt    5  18033.042 ±  188.738  ms/op
CachingClusteredClientCachingBenchmark.runQueries        noCache                10           1                          5                     1              minute        3             20000  avgt    5  49295.461 ±  452.770  ms/op
```
Can see more details about the benchmark in the benchmark test class. One thing I would like to point out is that it'll use two `ParallelMergeCombiningSequence` sequences when segment merged cache is enabled. 

##### Key changed/added classes in this PR
 * `CachingClusteredClient`
 * `CacheUtil`
 * `ForegroundCachePopulator`
 * `BackgroundCachePopulator`

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jihoonson** [comment]
Hi @yuanlihan, can you add some more details of the new feature, such as what it does, how it works, and how it is different from other caches? Ideally, this all should be documented in user docs. I think we can add docs as a follow-up of this PR.

**Comment by yuanlihan** [comment]
> Hi @yuanlihan, can you add some more details of the new feature, such as what it does, how it works, and how it is different from other caches? Ideally, this all should be documented in user docs. I think we can add docs as a follow-up of this PR.

sure, will document it later

**Review by asdf2014 — APPROVED** [review]
LGTM 👍

**Comment by jihoonson** [comment]
> It's expected as a better cache option for Broker in some use cases.

@yuanlihan perhaps I should have been clearer. More details of this new feature will clarify what problem you want to solve and help reviewers understand your intention better. For example, in your statement quoted above, "better" is quite ambiguous. you can explain in terms of what the new cache is better than others. You can add user-facing docs later.

**Comment by paul-rogers** [comment]
@yuanlihan, cool feature! Here is one question you might want to explain: what is that use case that will be improved by this change?

The feature would seem to optimize queries that hit exactly the same data every time. How common is this? It might solve a use case that you have, say, 100 dashboards, each of which will issue the same query. One of them runs the real query, 99 fetch from the cache.

Here's a similar use case I've seen many times, but would *not* benefit from a post-merge cache: a dashboard wants to show the "last 10 minutes" of data, and we have minute-grain cubes (segments). Every query shifts the time range by one minute which will add a bit more new data, and exclude a bit of old data. With this design, we cache the post-merge data, so each shift in the time window will result in a new result set to cache.

Can this feature instead cache pre-merge results? That way, we cache each minute slice once. The merge is still needed to gather up the slices required by the query. But, that merge is in-memory and should be pretty fast. The result would be, in this example, caching 1/10 the amount of data compared to caching post-merge data. And, less load on the historicals since we don't hit them each time the query time range shifts forward one minute.

In short, please explain your use case a bit more so we an understand the goal of this enhancement.

**Comment by yuanlihan** [comment]
Hi @paul-rogers, thanks for the comment.

>what is that use case that will be improved by this change?

This cache option can be useful for queries need to scan both historical and realtime segments and with time range filter, like the day so far or this week/month/year so far.

>The feature would seem to optimize queries that hit exactly the same data every time. How common is this? It might solve a use case that you have, say, 100 dashboards, each of which will issue the same query. One of them runs the real query, 99 fetch from the cache.

Let's say there's a dashboard with a today-so-far time range filter. It issues the same query to fetch realtime metric every minute. The cache feature would work like

1. At the moment, there are 10 historical segments and 2 realtime segments needed to be scanned. This cache feature will populate the merged result of the 10 historical segments to the cache. 
2. One minute later, as you described, the next query will fetch the merged result of the 10 historical segments and merge with result from real query. 
3. A few moments later, let's assume that another 2 new segments served on Historical, which means we need to scan 14 historical segments and 2 realtime segments. It'll prune the query and issue query for the other 2 historical segments and 2 realtime segments. Then it'll merge the previous cached result with the result of 2 new historical segments and populate the merged result of 14 historical segments to the cache. Last, it'll return the merged result of historical and realtime segments.
4. We can reuse the cached merged result, as long as the cached merged result contains the **_subset_** of historical segments needed to be scanned.

Is common? Hmm, It should be better if we have an efficient cache option for this use case.

>Here's a similar use case I've seen many times, but would not benefit from a post-merge cache: a dashboard wants to show the "last 10 minutes" of data, and we have minute-grain cubes (segments). Every query shifts the time range by one minute which will add a bit more new data, and exclude a bit of old data. With this design, we cache the post-merge data, so each shift in the time window will result in a new result set to cache.

The minute-grain segment sounds too strict to Druid. In my experience, the query should be fast if it needs to show "last 10 minutes" data since the last 10 minutes' data usually held  by peon workers. 
While this new cache feature should also be beneficial to fixed query with time filler, like last several hours or last one day/week. Let's assume that the query granularity is like `PT5M` for last-one-hour filter, `PT15M` for last-several-hours filter, and so on. And if the client is willing to align the start of the time interval filter based on the query granularity, then, every 5 minutes(or every 15 minutes), the fixed query will be like a query with fixed timestamp so far filter, which can potentially leverage the new cache feature.

>Can this feature instead cache pre-merge results? That way, we cache each minute slice once. The merge is still needed to gather up the slices required by the query. But, that merge is in-memory and should be pretty fast. The result would be, in this example, caching 1/10 the amount of data compared to caching post-merge data. And, less load on the historicals since we don't hit them each time the query time range shifts forward one minute.

As far as I observed, fetching a lot of cached entries from cache is not very efficient, especially for many large size cache entries. The pre-merge result caching capability you described sounds like the segment level cache on broker. While if the segment granularity is much larger than the query granularity, like with 1 hour segment granularity and 1 minute query granularity, then the segment level will not work. So I see what you mentioned. But I'm not sure whether it's efficient to have query granularity level slices caching feature. It's a nice point to discuss.

**Comment by jihoonson** [comment]
@yuanlihan thanks for the explanation. Your last comment is what I was looking for :slightly_smiling_face: I haven't read code yet, so perhaps my questions are very basic. So based on your comment, I suppose the problem you want to solve is that the current result-level cache could be not much useful if your query involves realtime segments? And the new post-merge cache can address it by caching only the merge results of historical results? This sounds reasonable to me. It would be nice if you can add a summarized version of the answers to these questions in the PR description. It will help others understand what problem you are solving.

> Can this feature instead cache pre-merge results? That way, we cache each minute slice once. The merge is still needed to gather up the slices required by the query. But, that merge is in-memory and should be pretty fast. The result would be, in this example, caching 1/10 the amount of data compared to caching post-merge data. And, less load on the historicals since we don't hit them each time the query time range shifts forward one minute.

This sounds like the segment-level cache on broker to me too. We already support it (https://druid.apache.org/docs/0.22.0/querying/using-caching.html#enabling-query-caching-on-brokers).

**Comment by yuanlihan** [comment]
> @yuanlihan thanks for the explanation. Your last comment is what I was looking for 🙂 I haven't read code yet, so perhaps my questions are very basic. So based on your comment, I suppose the problem you want to solve is that the current result-level cache could be not much useful if your query involves realtime segments? And the new post-merge cache can address it by caching only the merge results of historical results? This sounds reasonable to me. It would be nice if you can add a summarized version of the answers to these questions in the PR description. It will help others understand what problem you are solving.

Hi @jihoonson, thanks for looking into this.

I have updated the PR description for design review and your feedback will be appreciated.
